### PR TITLE
perf(cli): Optimize setting up `node_modules` on macOS

### PR DESCRIPTION
**Pull request by nathanwhit**
Hard linking (`linkat`) is ridiculously slow on mac. `copyfile` is better, but what's even faster is `clonefile`. It doesn't have the space savings that comes with hardlinking, but the performance difference is worth it imo.

```
❯ hyperfine -i -p 'rm -rf node_modules/' '../../d7/target/release/deno cache npm:@11ty/eleventy' 'deno cache npm:@11ty/eleventy'
Benchmark 1: ../../d7/target/release/deno cache npm:@11ty/eleventy
  Time (mean ± σ):     115.4 ms ±   1.2 ms    [User: 27.2 ms, System: 87.3 ms]
  Range (min … max):   113.7 ms … 117.5 ms    10 runs

Benchmark 2: deno cache npm:@11ty/eleventy
  Time (mean ± σ):     619.3 ms ±   6.4 ms    [User: 34.3 ms, System: 575.6 ms]
  Range (min … max):   612.2 ms … 633.3 ms    10 runs

Summary
  ../../d7/target/release/deno cache npm:@11ty/eleventy ran
    5.37 ± 0.08 times faster than deno cache npm:@11ty/eleventy
```

### CODE REVIEW

File: cli/util/fs.rs [hunk 1]
```diff
@@ -492,6 +492,68 @@ pub async fn remove_dir_all_if_exists(path: &Path) -> std::io::Result<()> {
   }
 }
 
+mod clone_dir_imp {
```
Comment (littledivy) [inline cli/util/fs.rs:495]: There is a similar optimization for `cp` and `copyFile` APIs. It uses rayon for recursive clonefile and has some other optimizations for small files and fallbacks: https://github.com/denoland/deno/blob/b21004b1d16ad7b67c7b1cd235abf792bf9d9777/ext/fs/std_fs.rs#L511

Maybe could be reused?

File: cli/util/fs.rs [hunk 2]
```diff
@@ -492,6 +492,68 @@ pub async fn remove_dir_all_if_exists(path: &Path) -> std::io::Result<()> {
...
+  use deno_core::error::AnyError;
+
+  use super::copy_dir_recursive;
+  use std::path::Path;
+
+  cfg_if::cfg_if! {
```
Comment (littledivy) [inline cli/util/fs.rs:501]: Nit (this isn't that bad without `cfg_if`):
```suggestion
#[cfg(target_vendor = "apple")]
mod apple {
  // ...
}

#[cfg(target_vendor = "apple")]
pub(super) fn clone_dir_recursive(from: &Path, to: &Path) -> Result<(), AnyError> {
  apple::clone_dir_resursive(from, to)
}

#[cfg(not(target_vendor = "apple"))]
pub(super) fn clone_dir_recursive(from: &Path, to: &Path) -> Result<(), AnyError> {
  // ...
}
```

**Review by littledivy — APPROVED** [review]
LGTM, just two comments

**Comment by lucacasonato** [comment]
@dsherret ptal if this looks good to you

File: cli/util/fs.rs (hunk 1 above)
Comment (dsherret) [reply cli/util/fs.rs:495]: I'm not sure we should use rayon in non-js sync specific scenarios. I feel like that might make stuff slower by using another thread pool.

**Review by dsherret — APPROVED** [review]
Nice!

> It doesn't have the space savings that comes with hardlinking, but the performance difference is worth it imo.

Agree. Storage is cheap.

File: cli/util/fs.rs (hunk 1 above)
Comment (nathanwhit) [reply cli/util/fs.rs:495]: Gave it a shot anyway, no performance difference but I found it caused a failure when the destination is a directory that already exists - didn't dig into it but seems like something wasn't getting copied over properly (the panic is from setting up the bin entries – the origin file we're trying to symlink doesn't exist)
```
thread 'main' panicked at cli/npm/managed/resolvers/local/bin_entries.rs:89:48:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

**Pull request merged**